### ISSUE TITLE
`interpolate` bicubic mode returns unexpected value on CUDA with float16 

### ISSUE THREAD

**Issue by yewentao256** [pytorch/pytorch#104157]
### 🐛 Describe the bug

Hello, I've encountered an issue while training a model that produces unusual output. Upon investigation, I discovered that the anomaly originates from an unexpected result from the 'interpolate' function.

```python
import torch
from torch.nn.functional import interpolate

def get_grads(dtype, size):
    input = torch.randn((2, 2, 1, 1), dtype=dtype, requires_grad=True).cuda()

    output = interpolate(input, size=size, mode='bicubic', align_corners=True)

    grad_outputs = [torch.ones_like(output)]
    grads = torch.autograd.grad([output], [input], grad_outputs)
    return grads[0]

print(get_grads(torch.float32, (128, 128)).flatten())
print(get_grads(torch.float16, (128, 128)).flatten())

print(get_grads(torch.float32, (64, 64)).flatten())
print(get_grads(torch.float16, (64, 64)).flatten())

print(get_grads(torch.float32, (32, 32)).flatten())
print(get_grads(torch.float16, (32, 32)).flatten())
```

The output is as follows:

```bash
tensor([16384., 16384., 16384., 16384.], device='cuda:0')
tensor([2048., 2048., 2048., 2048.], device='cuda:0', dtype=torch.float16)
tensor([4096., 4096., 4096., 4096.], device='cuda:0')
tensor([2048., 2048., 2048., 2048.], device='cuda:0', dtype=torch.float16)
tensor([1024., 1024., 1024., 1024.], device='cuda:0')
tensor([1024., 1024., 1024., 1024.], device='cuda:0', dtype=torch.float16)
```

Could this be a bug? It appears that the maximum value of the gradient is capped at 2048 for the float16 data type. If this is not a bug, could you please tell me why this is happening?

Furthermore, I'm wondering if you could implement a warning to alert users when this issue occurs.

### Versions

Collecting environment information...
PyTorch version: 2.0.0a0+gitc263bd4
Is debug build: False
CUDA used to build PyTorch: 11.2
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 7.5.0
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: glibc-2.17

Python version: 3.8.5 (default, Sep  4 2020, 07:30:14)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-3.10.0-693.el7.x86_64-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.2.67
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB
GPU 4: Tesla V100-SXM2-32GB
GPU 5: Tesla V100-SXM2-32GB
GPU 6: Tesla V100-SXM2-32GB
GPU 7: Tesla V100-SXM2-32GB

Nvidia driver version: 460.73.01
cuDNN version: /usr/local/cuda-9.0/lib64/libcudnn.so.7.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                48
On-line CPU(s) list:   0-47
Thread(s) per core:    2
Core(s) per socket:    12
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Gold 6146 CPU @ 3.20GHz
Stepping:              4
CPU MHz:               3200.000
BogoMIPS:              6400.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              25344K
NUMA node0 CPU(s):     0-11,24-35
NUMA node1 CPU(s):     12-23,36-47
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm arat pln pts

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.0.0a0+gitc263bd4
[pip3] torchvision==0.15.1a0
[conda] magma-cuda112             2.5.2                         1    https://pkg.sensetime.com/repository/conda-tsinghua-cloud/pytorch
[conda] mkl                       2023.1.0                 pypi_0    pypi
[conda] mkl-include               2023.1.0                 pypi_0    pypi
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.0.0a0+gitc263bd4          pypi_0    pypi
[conda] torchvision               0.15.1a0                 pypi_0    pypi

**Comment by lezcano** [comment]
cc @vfdev-5

**Comment by vfdev-5** [comment]
@yewentao256 this bug with float16 is related to this comment:
https://github.com/pytorch/pytorch/blob/9154bbc9990e9d024a6b613af6cda2f75ec6602e/aten/src/ATen/native/cuda/UpSample.cuh#L211-L215

When computing the gradients for `print(get_grads(torch.float16, (64, 64)).flatten())`,  we add one by one `+1` into grad input and due to float16 we can't accumulate after 2048:
```
s = torch.tensor(0.0, dtype=torch.float16)
for _ in range(64 * 64):
    s += torch.tensor(1.0, dtype=torch.float16)

print(s)
# 2048 instead of 4096
```
or 
```
s = torch.tensor(2048.0, dtype=torch.float16)
s += torch.tensor(1.0, dtype=torch.float16)
# 2048
```

So, it is a bug but following https://github.com/pytorch/pytorch/pull/19630#discussion_r281426912 : 
> maybe fix it later

**Comment by yewentao256** [comment]
@vfdev-5 Thank you very much! I understand why this could happen.

How about adding a warning in aten/src/ATen/native/cuda/UpSampleBicubic2d.cu to ensure that others don't have to debug when they encounter this issue?

Here's an example, function defined in `aten/src/ATen/native/cuda/UpSampleBicubic2d.cu`

```
static void upsample_bicubic2d_backward_out_cuda_template(
    const Tensor& grad_input,
    const Tensor& grad_output_,
    IntArrayRef output_size,
    IntArrayRef input_size,
    bool align_corners,
    c10::optional<double> scales_h,
    c10::optional<double> scales_w) {
  // Add a warning for float16 precision
  if (grad_output_.scalar_type() == at::ScalarType::Half) {
    TORCH_WARN_ONCE("Using float16 for the backward pass of upsample_bicubic2d can cause precision issues. If the result exceeds 2048, it will be truncated. For more details, refer to the discussion at https://github.com/pytorch/pytorch/issues/104157.");
  }

  // Rest of the function...
}
```

If this make sense, I'll make a pr for this issue.

**Comment by vfdev-5** [comment]
@yewentao256 I understand your point here however we have to be a bit more careful about the message and when it is applicable. I have an impression that all downsampling ops are not concerned with that and only upsampling ops (`upsample_*`, `grid_sample` and some others) can have a precision issue.

**Comment by yewentao256** [comment]
Yes, you are right. It appears that the `trilinear`, `bilinear`, `bicubic`, `area`, and `linear` modes of upsample share the same issues, with the exception of `nearest` and `nearest-exact`.

Let's first consider the upsample operation. If it makes sense to add warnings to these upsample operations, we can then contemplate extending these warnings to grid_sample and other operations that exhibit the same issue.

According to torch/nn/functional.py, it seems that the operation is dispatched to different kernels at this point:

```py
def interpolate(input: Tensor, size: Optional[int] = None, scale_factor: Optional[List[float]] = None, mode: str = 'nearest', align_corners: Optional[bool] = None, recompute_scale_factor: Optional[bool] = None, antialias: bool = False) -> Tensor:
    # ...
    if input.dim() == 5 and mode == "trilinear":
        assert align_corners is not None
        return torch._C._nn.upsample_trilinear3d(input, output_size, align_corners, scale_factors)
    if input.dim() == 4 and mode == "bicubic":
        assert align_corners is not None
        if antialias:
            return torch._C._nn._upsample_bicubic2d_aa(input, output_size, align_corners, scale_factors)
        return torch._C._nn.upsample_bicubic2d(input, output_size, align_corners, scale_factors)
    #...
```

What if we add the warnings at the Python layer? It could look something like this:

```py
def interpolate(input: Tensor, size: Optional[int] = None, scale_factor: Optional[List[float]] = None, mode: str = 'nearest', align_corners: Optional[bool] = None, recompute_scale_factor: Optional[bool] = None, antialias: bool = False) -> Tensor:  
    if has_torch_function_unary(input):
        return handle_torch_function(
            interpolate,
            (input,),
            input,
            size=size,
            scale_factor=scale_factor,
            mode=mode,
            align_corners=align_corners,
            recompute_scale_factor=recompute_scale_factor,
            antialias=antialias
        )

    if input.dtype == torch.float16 and mode in ('linear',' bilinear', 'bicubic', 'trilinear', 'area'):
        warnings.warn(f"Using float16 for the backward pass of upsample mode `{mode}` can cause precision issues. If the result exceeds 2048, it will be truncated. For more details, refer to the discussion at https://github.com/pytorch/pytorch/issues/104157.", UserWarning, stacklevel=2)
    # ...
```

**Comment by vfdev-5** [comment]
@yewentao256 I do not know what would be the best way to warn users about a precision issue.
Once again, I think this happens only on upsampling (in_size -> out_size > in_size) and not on downsampling (in_size -> out_size < in_size). The value 2048 is very specific to your example and thus the message lacks a bit of generality...

**Comment by lezcano** [comment]
If anything, this should probably go in the docs, not as a warning that anyone that uses the function gets.

**Comment by yewentao256** [comment]
@vfdev-5 Exactly, I didn't consider the downsampling cases, and I apologize for that oversight.

I fully agree with what @lezcano said. If it is challenging to identify the appropriate places to set the warning, it would be a good idea to include the information in the documentation. If someone encounters this issue and refers to the documentation, they will be able to resolve it easily.

So, would it look something like this?

```py
def interpolate(input: Tensor, size: Optional[int] = None, scale_factor: Optional[List[float]] = None, mode: str = 'nearest', align_corners: Optional[bool] = None, recompute_scale_factor: Optional[bool] = None, antialias: bool = False) -> Tensor:  # noqa: F811,B950
    r"""Down/up samples the input to either the given :attr:`size` or the given
    
    ...
    .. note::
        Mode ``mode='nearest-exact'`` ...
    .. note::
        **The content of our truncated issue here. **
    Note:
        {backward_reproducibility_note}
    """
```

**Comment by lezcano** [comment]
SGTM. Feel free to put up a PR and let's discuss it there.

**Comment by yewentao256** [comment]
> ```python
> 'linear',' bilinear', 'bicubic', 'trilinear', 'area'
> ```

Done

**Comment by bhack** [comment]
Are we affected by this also when using amp?
https://pytorch.org/tutorials/recipes/recipes/amp_recipe.html
